File: leetcode-68-Text_Justification/leetcode-68-Text_Justification.cpp
#include <iostream>
#include <cstdlib>
#include <vector>
#include <string>


std::vector<std::string> fullJustify(std::vector<std::string>& words, int maxWidth) {
    std::vector<std::string> res;

    std::vector<std::string> temp_res;
    size_t count_chars = 0;
    for(size_t i = 0; i<words.size(); ++i)
    {
        int temp_res_s = (temp_res.empty())? 0 : (temp_res.size()-1);
        int t = (count_chars+(words[i].size()+temp_res_s));
        if((t+1)<=maxWidth || (temp_res.empty()))
        {
            temp_res.emplace_back(words[i]);
            count_chars += (words[i].size());
        } else {
            int rem = (maxWidth-count_chars);
            int c_add = !(temp_res.size()-1)?0:(rem/(temp_res.size()-1));
            for(size_t j = 0; j<temp_res.size()-1; ++j)
            {
                temp_res[j].append(c_add,' ');
                count_chars += c_add;
            }
            rem = (maxWidth-count_chars);
            res.emplace_back("");
            size_t res_s = res.size();
            for(size_t j = 0; j<temp_res.size() || j<rem; ++j)
            {
                if(j<temp_res.size())
                {
                    res[res_s-1] += temp_res[j%temp_res.size()];
                }
                if(j<rem)
                {
                    res[res_s-1] += ' ';
                }
            }
            temp_res.clear();
            count_chars = 0;
            --i;
        }
        if(i==(words.size()-1))
        {
            int rem = (maxWidth-count_chars);
            res.emplace_back("");
            for(size_t j = 0; j<temp_res.size(); ++j)
            {
                res[res.size()-1] += temp_res[j];
                if(rem)
                {
                    res[res.size()-1] += ' ';
                } else {
                    break;
                }
                --rem;
            }
            res[res.size()-1].append(rem,' ');
        }
    }
    return res;
}


int main(){
    // std::vector<std::string> words = {"This", "is", "an", "example", "of", "text", "justification."};
    // int maxWidth = 16;

    // std::vector<std::string> words = {"Science","is","what","we","understand","well","enough","to","explain","to","a","computer.","Art","is","everything","else","we","do"};
    // int maxWidth = 20;

    // std::vector<std::string> words = {"What","must","be","acknowledgment","shall","be"};
    // int maxWidth = 16;

    std::vector<std::string> words = {"Listen","to","many,","speak","to","a","few."};
    int maxWidth = 6;

    auto res = fullJustify(words, maxWidth);
    
    std::cout << "[";
    for (size_t i = 0; i<res.size(); ++i)
    {
        std::cout << res[i];
        if(i!=(res.size()-1))
        {
            std::cout << ",";
        }
    }
    std::cout << "]" << std::endl;
    
    return EXIT_SUCCESS;
}

File: leetcode-68-Text_Justification/leetcode-68-Text_Justification.py
class Solution:
    def fullJustify(self, words: list[str], maxWidth: int) -> list[str]:
        result: list[str] = []
        line: list[str] = []
        line_len: int = 0
        word_i: int = 0
        while word_i < len(words):
            if (len(line) + line_len + len(words[word_i])) <= maxWidth:
                line.append(words[word_i])
                line_len += len(words[word_i])
                word_i += 1
            else:
                spaces_placeholders: int = len(line) - 1
                spaces_total: int = maxWidth - line_len
                if spaces_placeholders == 0:
                    line[len(line)-1] += ' ' * spaces_total
                else:
                    spaces_sharing: int = spaces_total // spaces_placeholders
                    for i in range(len(line)-1):
                        line[i] += ' ' * spaces_sharing
                        spaces_total -= spaces_sharing
                    if spaces_total != 0:
                        for i in range(len(line)-1):
                            if spaces_total == 0:
                                break
                            line[i] += ' '
                            spaces_total -= 1
                build_len_result: str = "".join(line)
                result.append(build_len_result)
                line = []
                line_len = 0
        else:
            spaces_total: int = maxWidth - line_len
            for i in range(len(line)-1):
                line[i] += ' '
                spaces_total -= 1
            if spaces_total != 0:
                line[len(line)-1] += ' ' * spaces_total
            build_len_result: str = "".join(line)
            result.append(build_len_result)
        return result


if __name__ == "__main__":
    words: list[str] = ["This", "is", "an", "example", "of", "text", "justification."]
    maxWidth: int = 16
    result = Solution().fullJustify(words, maxWidth)
    print(result)
